### DISCUSSION TITLE
StarryNightEffect MultiChannel Observation

### DISCUSSION THREAD

**Discussion by stretch911** [PlummersSoftwareLLC/NightDriverStrip#198] [General]
I've been extensively debugging effects due to micro-controller reboots...(which i am close to having conclusions on).

As a side effect of that debugging effort, I noticed on a 4 channel (4 strips) setup, the StarryNightEffect would "mostly" only do a single channel...highly disappointed, I started paying close attention. As posted in another thread, I added a debug statement to the main loop so I could track the current effect. I found that of all the "StarryNightEffects" I have in my effects.cpp, (there are several implementations of this effect as well as many "settings"), i found that one was working in multichannel... here is a sample of each...

       new StarryNightEffect<MusicStar>("Rainbow Music Stars", RainbowColors_p, 2.0, 2, LINEARBLEND, 5.0, 0.0, 20.0), // Rainbow Music Star - single channel
        new StarryNightEffect<MusicStar>("RGB Music Blend Stars", RGBColors_p, 0.8, 1, NOBLEND, 15.0, 0.1, 10.0),      // RGB Music Blur - Can You Hear Me Knockin' - multi channel

I found by tweaking the multichannel ("RGB Music Blend Stars"), the parameter that caused single vs multichannel is the "blurFactor" value. IF...you have blurFactor of 0.0 (which is default if you don't specify it), you only get a single channel. If you change that to .1 or .5, you suddenly get multichannel. I'm currently tracking down the "WHY?" and I do see some commented code, so will go back to history of this class "stareffect.h".

the effect really looks great w/ value of blurFactor of 0.0, but huge fail on single channel limitation...stay tuned.

**Comment by davepl** [comment]
Many original effects predate the multichannel support so only draw on the primary channel.  Later revs had it so that drawing to the default channel draws to ALL channels.

If you need an effect that needs a multi-channel Draw, and that has never had multichannel support before (ie: used in the atomic lamp or umbrella or similar effect) you’re likely the first person to do so, and may need to revise the code to respect the draw channel.

- Dave
